externals: Model as concrete specs and add dependency definitions

alalazo opened the pull request:
refers to #49697
closes #36179
closes #38638
closes #41682

fixes #26034
fixes #28827
fixes #6420  see https://github.com/spack/spack/issues/6420#issuecomment-720156534


In this PR externals are treated as concrete specs so there is a 1:1 mapping between an entry in `packages.yaml` and any installed external spec, for a fixed repository.

Their YAML specification has been extended, following #49697, to allow modeling dependencies of external specs. This might be quite useful to better capture e.g. ROCm installations that are already installed on a given system, or in similar cases. 

To be backward compatible with external specs specifying a compiler, for instance `mpich %gcc@9`, Spack will match the compiler specification to an existing external. It will fail when the specification is ambiguous, or leads to no match.



<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
Add spack.external to parse external specs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Start using external parser in compilers

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Allow ExternalSpecsParser to handle missing packages gracefully

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Treat externals as concrete specs in the solver

* Remove all the custom code paths dealing with externals
* Add the new SpecFilter.from_packages_yaml
* Fix unit tests

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Handle duplicate external IDs in ExternalSpecsParser

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Allow dependencies in external specs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add support for dependencies in the "packages" schema

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix bugs in architecture completion for external specs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Ensure variants are set as concrete when finalizing concretization

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Allow dealing with the old format `foo %gcc@X.Y` in external specs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix bootstrapping

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Find compilers before computing reused specs

Due to the nature of this change, we need to ensure
that compilers are found before we compute the reused
specs that are injected in the Solver setup phase.

Otherwise, in a situation where packages.yaml
doesn't exist, Spack will fail the first concretization,
and succeed in the following ones.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix unit test that is failing on macOS

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fix issue with Windows and reuse

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: account for target requirements in configuration

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Warn users when a dependency for externals is guessed

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Ensure we don't parse externals without concrete versions

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
solver: remove hack to add python deps to externals

Now Python dependencies can be specified in packages.yaml,
so remove the previous hack.

The feature should be niche enough that we don't need options
to recover the previous behavior.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
solver: add another option to complete external nodes

Allow users to configure external spec completion with
the new `concretizer:externals:completion` setting.

The current choices are to complete just the architecture
information or also the variants using default values.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
docs: document dependency modeling for externals

Add explanation and examples for specifying dependencies among
external packages in `packages.yaml`, including details on
`external_id`, `dependencies`, and `virtuals` attributes.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
docs: document completion options for external nodes

Explain the `concretizer:externals:completion` attribute
and its values, detailing how architectural information
and missing variants are handled for externals.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
spack find: add support for filtering external specs

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
spack find: add --show-configured-externals option

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add a unit test for default target requirements

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,318 @@
+            if not package_exists:
+                raise ValueError(f"Package '{node.name}' does not exist")
+
+            eid = external_dict.setdefault("external_id", str(uuid.uuid4()))
```
`uuid` is fine, these defaults are only necessary in memory to translate old specs like `mpich %gcc`

haampie commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -163,6 +161,55 @@ In addition, paths from ``extra_rpaths`` are added as library search paths for t
+       - spec: "mpileaks@2.3~debug+opt"
+         prefix: /user/path
+         dependencies:
+         - external_id: callpath_id
```
minor: if `deptypes` were required, then maybe

```
<id>:
  deptypes: link
```

to save a few characters and enforce the constraint of one edge between two nodes?

list is also fine, I guess. Just feels a bit verbose. Maybe `id` instead of `external_id` if the key is already under `externals` anyways?

alalazo commented:
> As for point (2), I propose we change how package hashes are calculated [ ... ]

I agree with the change of package hash, but I think it's orthogonal to this PR. I changed the description to reflect that there's a 1:1 mapping for a fixed package repo.

> I'm leaning towards setting no other variants on externals by default,  [ ... ]

I think that would be the preferred long term solution, but I would still make both strategies available right now in Spack. I'm also quite convinced that using default variants is probably the better default in the beginning, until we improve detection of packages. 

We already have "hidden" features, so I can move from the current config option into something like that.

psakievich commented on lib/spack/docs/build_settings.rst:
```diff
@@ -17,6 +17,25 @@ The default configuration is the following:
+
+This attribute currently allows two possible values:
+
+- ``architecture_only``: only the mandatory architectural information is completed on externals
+- ``default_variants``: external specs are also completed with missing variants, using their default values
```
Will users be able to set config values in the external spec and have them be respected? This is just default behavior, but it is it extendable based on user updates? From the implementation it seems so to me, so perhaps add this to the docs to explain how users can customize the specs?

One case that was reported to me recently was an `openssl` external which didn't resolve the `shared` variant correctly.  This triggered a bunch of rebuilds for in the python stack which wouldn't have been necessary if the user would have set the correct value for the variant.

alalazo replied:
> Will users be able to set config values in the external spec and have them be respected?

Yes. From the point of view of the configuration little changes. These are just behaviors to add missing pieces to the external spec. The difference is that in this PR this happens _before_ solving with clingo, and external specs are considered concrete.

psakievich replied:
My question is about composability of options/filters in the `find` command. There are a few that don't compose well with this command right now like `--format` and `--roots` IIRC.

I **think** it is not an issue though if this is all config based and not an install in the DB.

My motivating example is:

`spack -E find` vs `spack find`

The former is the way to see everything installed in the DB, and the later will only show what is available in the env. 

I would want the same feature set for environments to see what is only externals active for the environment. But if it all falls back to what is visible in the `packages` config scope then I think it is all taken care of naturally by changing the scopes seen with `spack -E find` vs `spack find`.  

Long winded way to double check.

psakievich replied:
Okay great. I think we should add a blurb about that in the docs saying users can set details about the spec and these options will fill in the rest if that isn't documented somewhere else.

psakievich commented on lib/spack/spack/cmd/find.py:
```diff
@@ -315,8 +328,12 @@ def root_decorator(spec, string):
-    concretized_but_not_installed = list()
-    if env:
+    concretized_but_not_installed = []
+    if args.show_configured_externals:
```
Is there a technical reason to separate this from the environment? I could see the case for externals the environment has a view to. Or is that taken care of already by config scopes?

alalazo replied:
> Is there a technical reason to separate this from the environment?

Yes. You can have externals in `packages.yaml` outside of an environment.

psakievich replied:
Sure. I suppose I'm mostly double checking that `spack -E find` and `spack find` with an active environment will still behave with the proper filters.  

I suppose if it is 100% config based and not dealing with installs in the DB then that filtering happens based on the config scopes though.

psakievich reviewed (commented):
I did a first pass here. Still working through tests. Only a couple of minor comments so far.

tgamblin commented on etc/spack/defaults/concretizer.yaml:
```diff
@@ -89,3 +89,10 @@ concretizer:
+    # Either 'architecture_only', to complete external specs with just the architecture of the
+    # current host, or 'default_variants' to complete external specs also with missing variants,
+    # using their default value.
+    completion: default_variants
```
It's weird to me that `architecture_only` and `default_variants` are not mutually exclusive policies, but they are the two mutually exclusive options here.

If architecture completion is always done, why not:

```yaml
externals:
    complete:       # arch is always completed
        - variants   # if not in the list, don't add them
        - # ... other option
        - # etc.
```

I don't have a super strong opinion here I just think the specification could be a little better.

alalazo replied:
What do we want to complete, on a node, besides architecture and variants?

tgamblin replied:
Version? Or do we require it already?  Seems like something that should really be provided or detected -- but I could imagine a user not knowing.

alalazo replied:
We require version already.

tgamblin commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -101,6 +101,55 @@ Specific limitations include:
+       - spec: "mpileaks@2.3~debug+opt"
+         prefix: /user/path
+         dependencies:
+         - external_id: callpath_id
```
`external_id` on a thing that's already in a list under `externals:` seems overly verbose -- how about just `id:`?  Same for dependencies.

alalazo replied:
Fixed in https://github.com/spack/spack/pull/51118/commits/49c145173b0f721fa93635ef3ebe3e8c40c196e7

tgamblin commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -101,6 +101,55 @@ Specific limitations include:
+         prefix: /user/path
+         external_id: mpich_id
+
+The ``external_id`` attribute is a unique string identifier of the external node being described.
+The ``dependencies`` attribute is a list of objects, where you can specify the ``external_id`` of the dependency and optionally the dependency types (``deptypes``), and the ``virtuals``, along the edge.
```
From the beginning, we've allowed people to refer to things by `spec` when possible.  I know *why* we need another id here, but in many cases we don't.  Note that because we are extending the existing externals schema, `id` is actually optional for most externals. You only really need an id if you are specifying multiple of something, and you may not need it then.  I do agree it can probably simplify the references in complicated scenarios.

I would default, though, to making the `id` the `spec`, and allowing the dependencies to refer to an external either by `spec:` or by `id:`.  I think in many cases the `spec` fields are going to be something like `mpich %gcc` or `mpich %intel` and the corresponding ids are going to be `mpich_gcc` and `mpich_intel`.  It seems redundant to force people to duplicate that information.

I also think that this somewhat simplifies your issue with converting entries like `mpich %gcc` to entries with dependencies... you can take the spec `gcc` and find a suitable external based on the spec -- something you're already kind of doing.  It makes a list of dependent externals pretty easy to specify:

```yaml
externals:
    - spec: "gcc@13.0.1"
    - spec: "gcc@12.0.0"
    - spec: "gcc@11.1.0"
    - spec: "intel-oneapi-compilers@2025.2.1"
    - spec: "mpich@3.1.1 %gcc@13.0.1"
    - spec: "mpich@3.1.1 %gcc@12.0.0"
    - spec: "mpich@3.1.1 %intel-oneapi-compilers@2025.2.1"
```

And you don't have to add much more than that.  I think an exact match is probably ok for a first cut here, but FWIW I think the minimal specification above could be something like:

```yaml
externals:
    - spec: "gcc@13.0.1"
    - spec: "gcc@12.0.0"
    - spec: "gcc@11.1.0"
    - spec: "intel-oneapi-compilers@2025.2.1"
    - spec: "mpich@3.1.1 %gcc@13"
    - spec: "mpich@3.1.1 %gcc@12"
    - spec: "mpich@3.1.1 %intel-oneapi-compilers"
```

i.e., you don't *have* to get into dependency specifications in the YAML unless it is necessary, and the YAML is interchangeable with spec syntax equivalents where they exist (something we've also done with toolchains).

tgamblin replied:
Note above I get that these would currently all have to in some structure like:

```yaml
gcc:
    externals:
        ...
intel-oneapi-compilers:
    externals:
        ...
mpich:
    externals:
        ...
```

but I also think it would not be awful to just start allowing something like:

```yaml
spack:
    externals:
        - ... # spec names make it obvious what these are for
    require:
        - spec: ...
          when: ...
    prefer:
        - spec: ...
          when: ...
```

which I think is easier to read than the overly nested and redundant `packages:` structure.  That's not for this PR though.

alalazo replied:
> That's not for this PR though.

Yeah, this should be another discussion (where I would concur for `externals`, but have to think about requirements because we would likely have a change in semantics if we do it like that)

alalazo replied:
**TL;DR** I don't think we should support `%` in `spec:` for externals going forward, as it leads to ambiguities and more duplication. The support for `mpich %gcc` etc. is there just to provide some backwards compatibility. 

I think we should _require_ an `id` if a spec has dependents, and work towards ways to have Spack add it to config when it detects externals with their dependencies. Matching by spec is error prone and will lead to issues.


---

Using an `id` in my opinion prevents us from having to deal with a lot of issues down the line, and a lot of guesses we have to make.

A few common operations done by users on externals include:

1. Updates to some of the specs manually
1. Deletion of some externals
1. Addition of new specs

These operations may be done on different configuration scopes.

### Scenario 1: require to use an `id`

Assume we enforce using `id`, for instance like this:
```yaml
mpich:
  externals:
  - spec: mpich@3.1.1
    prefix: ...
    dependencies:
    - id: XX
      deptypes: build

gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
    id: XX
  - spec: gcc@9.4.0 languages:='c,c++,fortran'
    prefix: ...
```
**Pros:**
1. We can update any of the specs without having to change anything else in the configuration. Dependency relationships are kept.
2. Likewise if we add a new spec to configuration
3. We can delete specs, and if we delete one `id` that was used we'll get an error when constructing the dependent.

### Scenario 2: use `spec` with an exact match

If we require users to have an exact match, then the situation is only worse:
```yaml
mpich:
  externals:
  - spec: mpich@3.1.1
    prefix: ...
    dependencies:
    - spec: gcc@14.2.0 languages:='c,c++,fortran'
      deptypes: build

gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
  - spec: gcc@9.4.0 languages:='c,c++,fortran'
    prefix: ...
```
**Cons:**
1. Any change to `gcc@14.2.0 languages:='c,c++,fortran'` must be duplicated in multiple places.[^1]
3. Specs for some externals can be quite long, so no gain on compactness of the specification by using the spec
4. Externals can be defined in different scopes. An update in a lower scope to `gcc` may break upstream configurations if we base it on exact match.

### Scenario 3: use `spec` with weaker semantics

If we use weaker semantics, we are in a situation with a lot of ambiguities:
```yaml
mpich:
  externals:
  - spec: mpich@3.1.1 %gcc
    prefix: ...

gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
  - spec: gcc@9.4.0 languages:='c,c++,fortran'
    prefix: ...
```
**Cons:**
1. Adding a spec may cause all `%gcc` externals to suddenly change, without erroring.[^2]
1. Removing a spec might do the same
1. Modifying a spec is also problematic, since the new spec might not match a previos `%<spec>` specification


[^1]: Here we have one, but in a real configuration it might be much more
[^2]: Think of adding `gcc@15.2.0`

tgamblin replied:
> **TL;DR** I don't think we should support `%` in `spec:` for externals going forward, as it leads to ambiguities and more duplication.

`id` IMO is duplication. I also think it may be necessary and/or convenient, but not all the time.

> The support for `mpich %gcc` etc. is there just to provide some backwards compatibility. 

This is fundamental to how people express themselves in config files -- it's how users interact with Spack.  The `id` adds confusion.  e.g., a user might ask, why is the id not a hash?  The answer to that takes a lot of explaining.  Probably more importantly , I don't think ids actually fix the problems you're saying they fix.

> ### Scenario 1: require to use an `id`
>
> Assume we enforce using `id`, for instance like this:
> 
> **Pros:**
> 1. We can update any of the specs without having to change anything else in the configuration. Dependency relationships are kept.

This isn't actually true. All this does is put the responsibility of generating a unique `id` on the user.  This is *exactly* the problem with manual installations the [original spack paper](https://tgamblin.github.io/pubs/spack-sc15.pdf) sought to address.  The implicit assumption here is that the user *knows* some unique id. In your example you didn't even come up with unique ids -- you used `XX`, because it's actually a pain to come up with unique ids 😄.

Suppose you had something slightly more ambitious than `XX`:

```yaml
mpich:
  externals:
  - spec: mpich@3.1.1
    prefix: ...
    dependencies:
    - id: gcc
      deptypes: build

gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
    id: gcc
```

Ok now add the second gcc:

```yaml diff
gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
    id: gcc
  - spec: gcc@9.4.0 languages:='c,c++,fortran'
    prefix: ...
    id: gcc
```

d'oh. I ran out of space in my id scheme.  Now I'm going to be lazy.  If I'm super lazy I'll just call the second `gcc` `gcc2`.  If I'm a little smarter, I'll call it `gcc9`.  Now my `gcc` id is inconsistently named, and maybe I'll need to disambiguate it some other way in the future. Or maybe I just leave it as `gcc` and accumulate some id tech debt.  I suspect this is the most likely thing.  Now we're back to the pre-spack software installation directory naming days: `gcc2`, `gcc3`, `gcc-todd`, `gcc-max`, `gcc-special`. I don't disagree that a set of unambiguous ids can help, but I also suspect that the only people generating them are going to be people who are not writing their YAML by hand.

The point of specs is to allow for a telescoping syntax where we *can* make sense of what the identifier means, and that there is a structured, regular way to construct ids -- one that the user doesn't have to think about. One that our users are already used to.

So:

> 3. We can delete specs, and if we delete one `id` that was used we'll get an error when constructing the dependent.

This *can be* true if you use specs as well.  IMO the problem is not the type of ids (specs are a type of id) but the fact that we might have an ambiguous specification. I agree that is a problem, and I think that we should *refuse* accept *any* ambiguous specification.  If we see one, we should warn the user.  We should do that regardless of whether we use ids or specs.

So, this (your example for 3):

```yaml
mpich:
  externals:
  - spec: mpich@3.1.1 %gcc
    prefix: ...

gcc:
  externals:
  - spec: gcc@14.2.0 languages:='c,c++,fortran'
    prefix: ...
  - spec: gcc@9.4.0 languages:='c,c++,fortran'
    prefix: ...
```

should lead to something like:

```console
==> Error: external for `mpich` is ambiguous. Which one do you mean?
    1.  spec: gcc@14.2.0 languages:='c,c++,fortran'
        prefix: ...
    2.  spec: gcc@9.4.0 languages:='c,c++,fortran'
        prefix: ...
```

We could also print a minimal disambiguating spec there if it helps, e.g.:

```console
==> Error: external for `mpich` is ambiguous. Which one do you mean?
    mpich %gcc@14
        spec: gcc@14.2.0 languages:='c,c++,fortran'
        prefix: ...
    mpich %gcc@9
        spec: gcc@9.4.0 languages:='c,c++,fortran'
        prefix: ...
```

For ids, this would look pretty much the same when the user provides a duplicate id.  In the case of both ids and specs, you have to warn the user when the scheme becomes ambiguous. It's worth noting that with Specs, you actually have a better way to narrow down what they might have been referring to, because specs inherently subselect. With an id you have no idea that the id even implies what the dependency package might be. Incidentally, that makes the file harder to read.  Or, if it does *not* make the file harder to read, then the user is encoding their ids as specs... which is redundant.

TL;DR: coming up with ids for a constantly changing set of things is hard -- but we built a way to do that in Spack -- it's Specs.  I think ids can be useful (e.g., for programmatically generated config) but I think you have the same ambiguity problems with ids and with specs, and ids put more burden -- and redundant work -- on the user.

I think we both agree we should not accept an ambiguous specification and that (2) above (exact matches) is bad.  So that is at least some progress 😄.

haampie replied:
This is pretty much a discussion about database normalization pros and cons :thinking:.

I think not mentioned yet is that a `spec` does not uniquely identify an external, cause the same `pkg` can be installed in `prefix: /foo` and `prefix: /bar`.

Todd convinced me though that it's annoying to require users to come up with a unique id since in general it's double bookkeeping.

The downside of `foo %bar` in externals is that it's lacking the deptype, though. Users still write `%bar` because they're used to Spack v0.

What about "Scenario 2" without the "exact match" behavior?

```yaml
packages:
  foo:
    externals:
    - spec: foo@1.2
      prefix: /usr
      dependencies:
      - spec: bar  # bar is sufficiently unique
        type: [link, run]
  bar:
    externals:
    - spec: bar@3.4 +x +y +z
      prefix: /usr
```

alalazo replied:
> What about "Scenario 2" without the "exact match" behavior?

That's what I would propose too. Basically, the only change I would make is to allow `spec:` in the `dependencies:` section of an external, instead of using an ID.

The top level external spec _should not_ contain externals. The only exception for now can be legacy specs specifying compilers, like `mpich %gcc`. This is what allows us to correctly guess the build-only type for the dependency. This exception can be removed in the future after a period of deprecation.

alalazo replied:
@tgamblin A note about this statement:

> I don't disagree that a set of unambiguous ids can help, but I also suspect that the only people generating one are going to be people who are not writing their YAML by hand.

I think this is basically true in the long term, but in a positive way.  The fact is, any use case that comes to my mind will involve enough dependencies that people will avoid writing edges in config themselves. This means using `spec:` over `id:` is likely to be the exception, not the rule.

Take for instance [the ROCm ecosystem](https://spack.readthedocs.io/en/latest/gpu_configuration.html):
```python
Spack version 1.1.0.dev0
Python 3.13.2, Linux x86_64
>>> from spack.concretize import concretize_one
>>> s = concretize_one("hip")
>>> len(s.dependencies())
22
```
Only the root has 22 dependencies, and if we don't take into account at least the relationships among the 7 nodes mentioned in the docs we will not be able to model the ROCm ecosystem in a satisfactory way. The number of edges is:
```python
>>> len(list(s.traverse(cover="edges")))
752
```
Say that we need to model one order of magnitude less for those 7 dependencies. I don't think there are users that will regularly write ~70 edges manually for ROCm, including their dependency types. What I think is more likely is that they expect:
```console
$ spack external find hip
```
to write those dependencies for them.

alalazo replied:
@tgamblin See if the last commit looks better

tgamblin replied:
> I think this is basically true in the long term, but in a positive way. The fact is, any use case that comes to my mind will involve enough dependencies that people will avoid writing edges in config themselves. This means using spec: over id: is likely to be the exception, not the rule.

I don't disagree w/this in the long term, but in the short (and for small stuff) I think it's very useful to have an intuitive shorthand for this stuff -- one that doesn't make you wince when you have to throw something together to test, debug, or just try something. There are cases (e.g. things that depend on python) where adding just the one dependency can be quite helpful. Python externals are basically auto-wired in a somewhat dubious way because they're not very useful without that. I think making that more obvious in config is good (and would also allow us to specify the python dependency more accurately in non-unified environments).

I don't think auto-detecting wiring, at least with RPATHs, is that hard. If you know where to find the dependency library you can figure out what dependencies can be wired together. For module systems and other things that rely on LD_LIBRARY_PATH I think it *is* kind of hard... or at least it merits more thought. There are probably some general heuristics that would work well for, say, basic RPM-based ROCm installs, system packages, and things like EESSI where they do use RPATHs... all things I would like us to auto-detect well.

tgamblin reviewed (changes requested):
Generally this is great. 

Some high level requests for now -- will go through in more detail but I don't expect to find too much to gripe about in the lower level implementation.

alalazo committed:
Shorten "external_id" to "id"

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Allow specifying dependencies by spec:

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

tgamblin commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -102,6 +102,90 @@ Specific limitations include:
+     mpich:
+       externals:
+       - spec: "mpich@3.0.4"
+         prefix: /user/path
```
Another example with two `mpich`'s or something to show a potentially not-unique case and how to disambiguate would be useful.

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -988,59 +974,13 @@ possible_provider_weight(node(ProviderID, Provider), VirtualNode, 100, "fallback
-% if a package is not buildable, only externals or hashed specs are allowed
-external(node(ID, Package))
+% if a package is not buildable, only concrete specs are allowed
+error(100, "Cannot build {0}, since it's marked non-buildable and no externals satisfy the requests", Package)
```
```suggestion
error(100, "Cannot build {0}, since it is configured `buildable:false` and no externals satisfy the request", Package)
```

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -575,6 +563,26 @@ attr("concrete_variant_set", node(X, A1), Variant, Value, ID)
+  has_built_packages().
+
+
+
```
```suggestion


```

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -575,6 +563,26 @@ attr("concrete_variant_set", node(X, A1), Variant, Value, ID)
+  compiler_from_reuse(Hash, DependencyPackage),
+  not node_compiler(_, node(_, DependencyPackage)),
+  % We don't want to give penalties if we're just installing binaries
+  has_built_packages().
```
I think `will_build_packages()` makes the wording clearer...

tgamblin reviewed (changes requested):
partial review with some small changes -- mostly LGTM so far.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
```
Needs a docstring saying this is our internal rep for externals, describing how it's used.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+
+
+class ExternalDict(TypedDict, total=False):
+    """Dictionary representation of an external spec."""
```
should mention that this is meant to match packages.yaml schema

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def complete_architecture(node: spack.spec.Spec) -> None:
+    """Completes a node with architecture information."""
```
should say how it guesses

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def complete_variants_and_architecture(node: spack.spec.Spec) -> None:
+    """Completes a node with variants and architecture information."""
```
should say a bit more about how

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                if not node.satisfies(f"{v.name}=*"):
+                    # Cannot use Spec.constrain, because we lose information on the variant type
+                    node.variants[v.name] = v.make_default()
+            progress = True
```
Minor tweaks to make this clearer:
```suggestion
    variants_dict = pkg_class.variants.copy()
    changed = True
    
    while variants_dict and changed:
        changed = False
        items = list(variants_dict.items())  # copy b/c loop modifies dict
        
        for when, variants_by_name in items:
            if not node.satisfies(when):
                continue
            variants_dict.pop(when)
            for name, vdef in variants_by_name.items():
                if name not in node.variants:
                    # Cannot use Spec.constrain, because we lose information on the variant type
                    node.variants[name] = vdef.make_default()
            changed = True
```

I don't think this is strictly necessary for correctness:

```python
    while variants_dict and changed:
```

but it makes it clearer what the loop is doing, and it avoids a needless last iteration (and copy) if you exhaust variants_dict *and* change something at the same time.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def extract_dicts_from_configuration(packages_yaml) -> List[ExternalDict]:
+    """Extracts external specs from a configuration dictionary."""
```
How? Should say something about precedence of `"all"` here. Mention how it adds target info.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+    extra_attributes: Dict[str, Any]
+    id: str
+    dependencies: List[DependencyDict]
+    required_target: str
```
add a comment to note that this one deviates fro the YAML (it's special)

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+_TARGET_RE = re.compile(r"target=([^\s:]+)")
+
+
+def _required_target(entry) -> str:
```
docstring

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+    config: ExternalDict
+
+
+class ExternalSpecsParser:
```
docstring

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+        complete_node: Callable[[spack.spec.Spec], None] = complete_variants_and_architecture,
+        allow_nonexisting: bool = True,
+    ):
+        """Initializes a class to manage and process external specifications.
```
example external yaml might be nice here. Also should mention that we're talking about externals in `packages.yaml`

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                warnings.warn(
+                    f"cannot constrain external spec '{external_dict['spec']}' with target "
+                    f"'{external_dict['required_target']}'. This spec will not be considered "
+                    f"during concretization."
```
I think this can result in a lot of warnings... Can this be debug?  If we document that only externals with satisfying targets will be considered in a solve, and maybe later add something to show what externals were considered in a solve, I think it's sufficient. People with multi-platform ends are going to hate this one.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                )
+                continue
+            except ExternalSpecError as e:
+                warnings.warn(f"{e} Fix your packages.yaml configuration.")
```
Will this print the location in the file where the error is?  Config errors should all be striving to do that -- users complain about this all the time and the dictionary from config should have marks on it telling what lines the offending entry was on.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                raise spack.repo.UnknownPackageError(node.name, repo=spack.repo.PATH)
+
+            if not package_exists:
+                raise ValueError(f"Package '{node.name}' does not exist")
```
Last condition here is unreachable.

```suggestion
            if not package_exists:
                if self.allow_nonexisting:
                    continue
                raise spack.repo.UnknownPackageError(node.name, repo=spack.repo.PATH)
```

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+            if not package_exists:
+                raise ValueError(f"Package '{node.name}' does not exist")
+
+            eid = external_dict.setdefault("id", str(uuid.uuid4()))
```
I think it'll be easier and faster to debug this if you make a reserved id with a counter instead, e.g. `_spack_external_id_{n}`

alalazo replied:
Not sure if I agree on this one. I'd have to check if we used a number that was already taken by the user, or put a restriction on user-defined id. Also, I was thinking of using uuid4 in a later PR for detection, and persist them.

tgamblin replied:
I guess we can always change this.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                raise DuplicateExternalError(
+                    f"Specs {node} and {other_node} have the same external id {eid}."
+                    f" Fix your packages.yaml configuration."
+                )
```
this should also display clear file and line info -- also "Fix your packages.yaml configuration" is a little too... adversarial?  I think just showing the error and saying that ids need to be unique is fine.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                current_node._add_dependency(dependency_node, depflag=depflag, virtuals=virtuals)
+
+        for node in self.nodes:
+            node._finalize_concretization()
```
I think it is worth breaking this out of `_parse`. I don't care if `_parse` calls it, but I think it should be more than a single method.

alalazo replied:
Split `_parse` into multiple methods in https://github.com/spack/spack/pull/51118/commits/a06fdb3fe4851a864504f0893899d70ef8328a0f

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+            current_node = entry.spec
+            current_dict = entry.config
+            is_legacy = False
+            # Guess how to convert old entries like 'mpich %gcc' to a dependency in the dict
```
This isn't old, right? It's supported.

alalazo replied:
The idea is to support only `%<compiler>` in the root spec for backward compatibility. The code to support more syntax is there, but I think it would be confusing to have different ways to refer to dependencies because of the defaults for dependency types. 

I guess we both agree that mixing the two ways to refer to dependencies should not be allowed:
```yaml
spec: foo%bar
dependencies:
- spec: baz
```
because it's just increasing the complexity of the code base for no reason (we'd have to check whether a dependency has been mentioned in multiple ways etc.).

If we assume only one way is used for a given external, we still have inconsistencies in the defaults which are extremely confusing imo. For instance:
```yaml
spec: foo %[deptypes=build,link] bar
```
would be equivalent to:
```yaml
spec: foo
dependencies:
- spec: bar
  deptypes: [build,link]
```
but:
```yaml
spec: foo %bar
```
_is not equivalent_ to
```yaml
spec: foo
dependencies:
- spec: bar
```
The dependency `bar` is assumed to be a build only dependency in the first case (since the only legacy case for it are compilers), while it's a `build,link` dependency in the second case (to be consistent with the default of `depends_on`)

tgamblin replied:
I still don't think we should disallow using both types of syntax. It is intuitive to write specs, we can express things with specs -- I think it's much more confusing if a user writes a spec and Spack doesn't understand it.

Per our discussion, let's keep both. I don't mind if we disallow `foo %bar` with `dependencies:spec:bar`, but I think you should be able to write `foo %bar` with `dependencies:spec:baz`. 

On how to resolve the inconsistency, I think:
1. `%` should be interpreted based on the Package like it is everywhere else (i.e. as concretization would do)
2. You can infer that it's a build dep if all possible deps on it in a package are build dependencies, e.g. something like:
    ```python
    cls = spack.repo.PATH.get_pkg_class("hdf5")
    spec = Spec("gcc")
    spec_is_build_only = all(
        dep.depflag == dt.BUILD
        for when, by_name in cls.dependencies.items() for name, dep in by_name.items()
        if spec.intersects(dep.spec)
    )
    ```
    if `spec_is_build_only` is True, make it a build dep, otherwise make it a build/link dep. Maybe we can do better with more analysis later, but I think this is ok for now.

alalazo replied:
I went further and inferred the `deptype` when not specified

alalazo replied:
> I still don't think we should disallow using both types of syntax.

Currently it's either one syntax, or the other. If you agree, and people ask for it, we can allow both together later.

tgamblin replied:
Yeah leave it for now and let’s allow both later.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    raise ExternalDependencyError(
+                        f"The external spec '{current_dict['spec']}' has an invalid dependency"
+                        f" specification. Fix your packages.yaml configuration."
+                    )
```
we should accept this as though they provided `deptypes:` in yaml -- I see no difference.

alalazo replied:
This is the same discussion as https://github.com/spack/spack/pull/51118#discussion_r2429319583

alalazo replied:
Done

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    raise ExternalDependencyError(
+                        f"The external spec '{current_dict['spec']}' depends on "
+                        f"'{edge.spec.name}', but there is no such external spec in packages.yaml."
+                    )
```
file and line info for the first spec here.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    raise ExternalDependencyError(
+                        f"The external spec '{current_dict['spec']}' depends on '{edge.spec}',"
+                        f" but there is no '{edge.spec.name}' that satisfies the request "
+                        f"in packages.yaml."
```
file and line info

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                warnings.warn(
+                    f"the external spec '{current_dict['spec']}' has been guessed to depend on "
+                    f"'{selected.config['spec']}'. If this is incorrect, fix your packages.yaml."
+                )
```
This is guessing for compilers right?  I think this should fail just like other externals with deps if the spec is ambiguous. Can we not make `spec: mpich %gcc` the same code path as `externals:[{dependencies:spec:gcc}]`?

alalazo replied:
> Can we not make spec: mpich %gcc the same code path as externals:[{dependencies:spec:gcc}]?

Done in https://github.com/spack/spack/pull/51118/commits/ea384cda04c017a21e8fe75c2c14f93b9edd2527.

Note that this will cause some failures (see how the unit tests had to change). The fixes to the failures are backward compatible, and can be used with older Spack versions.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    continue
+
+                if "spec" not in dependency_dict:
+                    raise ExternalDependencyError(
+                        f"the spec {current_dict['spec']} needs to specify either the id or the "
+                        f"spec of its dependencies."
+                    )
```
this is going to be really hard to understand without file and line info.

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    raise ExternalDependencyError(
+                        f"A dependency for {current_dict['spec']} has an external id "
+                        f"{dependency_id} that is not defined in packages.yaml"
+                    )
```
file / line info

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,349 @@
+                    )
+
+                query_spec = spack.spec.Spec(dependency_dict["spec"])
+                candidates = [
+                    x for x in self.specs_by_name[query_spec.name] if x.spec.satisfies(query_spec)
+                ]
+                if len(candidates) == 0:
+                    raise ExternalDependencyError(
+                        f"the spec {current_dict['spec']} depends on {query_spec}, but there is "
+                        f"no such external spec in packages.yaml."
+                    )
+                elif len(candidates) > 1:
+                    raise ExternalDependencyError(
+                        f"the spec {current_dict['spec']} depends on {query_spec}, but there are "
+                        f"multiple external specs in packages.yaml that satisfy it."
+                    )
```
Looks good but needs file / line info. Would be nice if errors could be accumulated and output all at once at the end instead of fail-fast like this (so you see them all) but this is fine for 1st cut.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2222,99 +2180,11 @@ def external_packages(self):
-            if not external_buildable or externals:
-                self.gen.h2(f"External package: {pkg_name}")
-
             if not external_buildable:
```
Can be inlined at this point:

```suggestion
            if not data.get("buildable", True):
```

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -4032,100 +3863,6 @@ def _specs_with_commits(spec):
-    """
-    Ensure all external python packages have a python dependency
-
-    If another package in the DAG depends on python, we use that
-    python for the dependency of the external. If not, we assume
-    that the external PythonPackage is installed into the same
-    directory as the python it depends on.
-    """
-    # TODO: Include this in the solve, rather than instantiating post-concretization
-    if "python" not in dependent_package.spec:
-        if extendee_spec:
-            python = extendee_spec
-        else:
-            python = _get_external_python_for_prefix(dependent_package)
-            if not python.concrete:
-                repo = spack.repo.PATH.repo_for_pkg(python)
-                python.namespace = repo.namespace
-
-                # Ensure architecture information is present
-                if not python.architecture:
-                    host_platform = spack.platforms.host()
-                    host_os = host_platform.default_operating_system()
-                    host_target = host_platform.default_target()
-                    python.architecture = spack.spec.ArchSpec(
-                        (str(host_platform), str(host_os), str(host_target))
-                    )
-                else:
-                    if not python.architecture.platform:
-                        python.architecture.platform = spack.platforms.host()
-                    platform = spack.platforms.by_name(python.architecture.platform)
-                    if not python.architecture.os:
-                        python.architecture.os = platform.default_operating_system()
-                    if not python.architecture.target:
-                        python.architecture.target = spack.vendor.archspec.cpu.host().family.name
-
-                python.external_path = dependent_package.spec.external_path
-                python._mark_concrete()
-        dependent_package.spec.add_dependency_edge(
-            python, depflag=dt.BUILD | dt.LINK | dt.RUN, virtuals=()
-        )
```
I don't think we can just remove this as it's a behavior people rely on -- how is this behavior handled now?  Are users required to hook up python packages to python?  We should think about what the path of least pain is here... I think if we don't want to break old envs, we need to grandfather some part of this auto-wiring in. 

same for `_get_external_python_for_prefix` below -- what's the migration strategy?

alalazo replied:
See https://github.com/spack/spack/pull/51118/commits/160f7afe1acb4087b7547cad2911cee3c01501ab This recovers minimal support. If we have a python external, we'll add the dependency automatically. Otherwise we'll fail like this:
<img width="3840" height="269" alt="Screenshot from 2025-10-16 21-59-35" src="https://github.com/user-attachments/assets/4e45e422-3f0a-4212-a429-325fd126c1ae" />

tgamblin commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -102,6 +102,90 @@ Specific limitations include:
 * The logic does not search through module files, it can only detect packages with executables defined in ``PATH``; you can help Spack locate externals which use module files by loading any associated modules for packages that you want Spack to know about before running ``spack external find``.
 * Spack does not overwrite existing entries in the package configuration: If there is an external defined for a spec at any configuration scope, then Spack will not add a new external entry (``spack config blame packages`` can help locate all external entries).
 
+Specifying dependencies among external packages
```
We might want to think about the order of this part of the manual, but I think this should go *after* we describe `buildable: false` -- I guess I think dependencies are more advanced -- do people agree?

tgamblin commented on lib/spack/docs/packages_yaml.rst:
```diff
@@ -102,6 +102,90 @@ Specific limitations include:
+         prefix: /user/path
+         id: mpich_id
+
 Prevent packages from being built from sources
```
I am not sure whether it goes in here, but somewhere we should be clear about the priority of reuse for externals -- basically what a user should expect about whether their external will be used.  e.g., I don't think it's always obvious that:

1. If you have an external and an internal version of something, which one will be preferred?
2. What if they're the same version?
3. How does `buildable: false` mean w.r.t. reuse preferences?
4. A pointer to `concretizer:reuse:` preferences is probably in order somewhere in here so that users get the gory details if they need them -- that's not present currently.

alalazo committed:
Fix unit tests after 51417

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Inline local variable

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
has_built_packages -> will_build_packages

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Improve error message

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Improve docstrings + minor refactors

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Demote a warning to debug print

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Attach line information to builtin dicts from YAML

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Split ExternalSpecParser._parse into multiple methods

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Add line information to error, warning and debug prints

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: use the same code path for `foo%bar` and externals specifying `dependencies:`

 This may cause some backwards incompatibilities, but Spack should
 report clear errors for ambiguous specs and the fixes in configuration
 files are backward compatible.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: add Python dependencies automatically

This recovers part of the automatic Python dependency support.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: allow specifying inline deptypes and virtuals

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: infer deptypes for inline dependencies

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: ensure inline and YAML syntax are equivalent

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
externals: add module docstring

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Improve documentation

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

tgamblin commented on lib/spack/spack/externals.py:
```diff
@@ -0,0 +1,425 @@
+                and not node.dependencies()
+                and any([c.__name__ == "PythonExtension" for c in pkg_class.__mro__])
+            ):
+                warnings.warn(
+                    f"Spack is trying attach a Python dependency to '{node}'. This feature is "
+                    f"deprecated, and will be removed in v1.2. Please make the dependency "
+                    f"explicit in your configuration."
+                )
```
I think this should say something more specific and tell them what to do, with file and line info, but it can be in some later PR.  Something like:

```suggestion
                warnings.warn(
                    f"{file}:{line}: Python extension {node} does not define a python dependency is "
                    f"guessing that it depends on {python_node} at {line_info}. Spack will stop guessing "
                    f"this dependency in version 1.2 and you will need to specify it explicitly. You can add "
                    f"the following to your `packages.yaml` to do this:"
                    <python extension YAML augmented with dep with suitably unique python spec>
                )
```

I actually think 1.2 is too soon to remove this -- I think we need a better solution (e.g. with detection or some kind of warning when packages that really need a particular type of dependency don't have it).  I think we should come up with something in the package API, e.g., make extensions require their extendee as a dep in `packages.yaml` AND do a good job of detecting it, before we remove this.

tgamblin reviewed (approved):
Assuming the coverage is there, I think this is great! Thanks for the iteration and for addressing most of the points.

I think the error messages and upgrade paths need work before the release but we have time to improve those.  See comment above but don't let I hold up merging.
